content/blogposts/2023/building-conc-better-structured-concurrency-for-go.md

```diff
@@ -0,0 +1,326 @@
+To me, the API feels very natural now and most of the tweaks I did were to extend the goal of making it hard to use incorrectly. From my experience with other libraries, and looking at other concurrency implementations for inspiration, I found that nearly all of them suffer from at least one of the following problems.
+
+* They feel **unnatural** to use 
+* They contain many **foot guns** or they make it easy to do dangerous things without realizing
+* They make it **hard to understand** what's actually happening under the hood
+* They contain **surprising** implementation details that catch users off guard
+
+Often I found that I had to reduce the feature set of Conc in order to keep it simple and safe. I've made some slight tweaks after getting some feedback from users that further these goals. 
+
+For example, at Sourcegraph we generally use `GO_MAX_PROCS` to limit Goroutines to the number of core available. For us, that's often a sane default, but Go's scheduler can happily spawn 1000s of routines. For a majority of users, having this somewhat arbitrary limit was **surprising**, so by following the goal of "follow the principle of least surprise" it made sense to remove this.
+
+### challenge #2: Discovering edge cases while handling panics
+
+Panics aren't meant to be used as a way of exception handling, but while writing `conc` I needed to be able to catch and handle panics. One thing I learned about Go in spite of coding in it for years is that when you intercept a panic with a `nil` value, it will look like nothing actually went wrong.
+
+For example, running hte following code will print "did not panic", even though a panic occurs.
+
+```go
+package main
+
+func main() {
+	defer func() {
+		if r := recover(); r != nil {
+			println("caught a panic!")
+		} else {
+			println("did not panic")
+		}
+	}()
+
+	panic(nil)
+}
+```
+
+## Releasing `conc` to the world
+
+After I made the GitHub repository public, I posted about `conc` internally. I expected my coworkers to look over it and maybe give some comments or criticism, but it quickly got shared over [to Reddit](https://www.reddit.com/r/golang/comments/1022agb/conc_better_structured_concurrency_for_go/), where it received a bit of (mainly positive) attention. There's since been some discussion [on GitHub](https://github.com/sourcegraph/conc/issues?q=is%3Aissue+is%3Aclosed) and [on Hacker News](https://news.ycombinator.com/item?id=34344514) and I've made some small tweaks based on feedback.
+
+It's been fun to watch `conc` spreading across GitHub as people discover it. Currently there are [423 matches](https://sourcegraph.com/search?q=context%3Aglobal+sourcegraph%2Fconc&patternType=standard&sm=1&groupBy=repo) for `sourcegraph/conc`. Many of these are automatic summaries of trending repositories or similar, but it's also made its way for example into [Kong](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller/-/blob/internal/dataplane/kong_client.go) already.
```

Comment:
I'm a little uncomfortable specifically calling out projects that are using it, and the sourcegraph search doesn't really provide much interesting info other than "this appeared on a lot of lists." I think I'd prefer we drop this paragraph